File: app/assets/javascripts/student_profile/transitionNoteParser.js
const STRENGTHS_PROMPT = `What are this student's strengths?`;
const COMMUNITY_PROMPT = `What is this student's involvement in the school community like?`;
const PEERS_PROMPT = `How does this student relate to their peers?`;
const GUARDIAN_PROMPT = `Who is the student's primary guardian?`;
const OTHER_PROMPT = `Any additional comments or good things to know about this student?`;


function escapeToUseAsRegexLiteral(x) {
  return x.replace('?', '\\?');
}

function clean(text) {
  return text.trim().replace(/[\—\-\_]/g, '').trim();
}

function extractOne(text, beforeText, afterText = undefined) {
  const beforeRegexLiteral = escapeToUseAsRegexLiteral(beforeText);
  const afterRegexLiteral = (afterText)
    ? escapeToUseAsRegexLiteral(afterText)
    : '';
  const regex = new RegExp(beforeRegexLiteral + '([\\s\\S]*)' + afterRegexLiteral);
  const matches = text.match(regex);
  if (!matches) return null;
  return clean(matches[1]);
}

// See transition_note_parser.rb
function parseTransitionNoteText(transitionNoteText) {
  return {
    strengths: extractOne(transitionNoteText, STRENGTHS_PROMPT, COMMUNITY_PROMPT),
    community: extractOne(transitionNoteText, COMMUNITY_PROMPT, PEERS_PROMPT),
    peers: extractOne(transitionNoteText, PEERS_PROMPT, GUARDIAN_PROMPT),
    guardian: extractOne(transitionNoteText, GUARDIAN_PROMPT, OTHER_PROMPT),
    other: extractOne(transitionNoteText, OTHER_PROMPT)
  };
}

// Parse and re-render as text with whitespace etc. in a standard format.
export function parseAndReRender(transitionNoteText) {
  const {strengths, community, peers, guardian, other} = parseTransitionNoteText(transitionNoteText);
  const NEWLINE = "\n";
  return [
    STRENGTHS_PROMPT, NEWLINE, strengths, NEWLINE, NEWLINE,
    COMMUNITY_PROMPT, NEWLINE, community, NEWLINE, NEWLINE,
    PEERS_PROMPT, NEWLINE, peers, NEWLINE, NEWLINE,
    GUARDIAN_PROMPT, NEWLINE, guardian, NEWLINE, NEWLINE,
    OTHER_PROMPT, NEWLINE, other, NEWLINE, NEWLINE
  ].join('');
}

File: app/lib/transition_note_parser.rb
class TransitionNoteParser
  STRENGTHS_PROMPT = "What are this student's strengths?"
  COMMUNITY_PROMPT = "What is this student's involvement in the school community like?"
  PEERS_PROMPT = "How does this student relate to their peers?"
  GUARDIAN_PROMPT = "Who is the student's primary guardian?"
  OTHER_PROMPT = "Any additional comments or good things to know about this student?"

  def parse_text(text)
    {
      strengths: extract_one(text, STRENGTHS_PROMPT, COMMUNITY_PROMPT),
      community: extract_one(text, COMMUNITY_PROMPT, PEERS_PROMPT),
      peers: extract_one(text, PEERS_PROMPT, GUARDIAN_PROMPT),
      guardian: extract_one(text, GUARDIAN_PROMPT, OTHER_PROMPT),
      other: extract_one(text, OTHER_PROMPT)
    }
  end

  private
  def extract_one(text, before_text, after_text = nil)
    before_regex_literal = escape_to_use_as_regex_literal(before_text)
    after_regex_literal = escape_to_use_as_regex_literal(after_text)
    regex = Regexp.new(before_regex_literal + '([\\s\\S]*)' + after_regex_literal)
    match_text = regex.match(text).captures[0]
    clean(match_text)
  end

  def escape_to_use_as_regex_literal(text)
    return '' if text.nil?
    Regexp.escape(text)
  end

  def clean(text)
    text.strip.gsub('[\—\-\_]', '')
  end
end
